R/MsBackendTimsTof-functions.R

```diff
@@ -88,4 +88,35 @@
 
 MsBackendTimsTof <- function() {
   new("MsBackendTimsTof")
-}
\ No newline at end of file
+}
+
+#' Read peaks from a single .d folder.
+#'
+#' @param x `character(1)` indicating the file to read from.
+#'
+#' @noRd
+# .timstof_peaks <- function(x) {
+#   tms <- OpenTIMS(x)
+#   tmp <- query(tms, tms@frames$Id, c("frame", "scan", "mz", "intensity"))
+#   # the result below should be a list of matrices (and not data.frames), right?
+#   unname(split.data.frame(as.matrix(tmp[, c("mz", "intensity")]), 
+#                           factor(paste(tmp$frame, tmp$scan))))
+# }
+
+#' Read columns (mz, intensity or both) from a single file
+#' 
+#' @param x `character(1)` indicating the file to read from.
+#'
+#' @noRd
+.read_frame_col <- function(x, columns) {
+  tms <- OpenTIMS(x)
+  if (any(!columns %in% tms@all_columns))
+    stop("Invalid value for columns")
+  if (!length(sd <-setdiff(columns, c("frame", "scan"))))
+    stop("At least one column value different from 'frame' and 'scan' required")
+  tmp <- query(tms, tms@frames$Id, c("frame", "scan", sd))
+  f <- factor(paste(tmp$frame, tmp$scan))
```

Comment:
before defining the `factor` I suggst to eventually subset `tmp` based on the values in the `"scans"` columns of the indices matrix.

Reply:
The `paste` and `factor` calls here are the performance bottlenecks. Maybe try also an alternative way how to extract the values from this `data.frame`? Maybe even using a loop (you know the frames and scans you want from the indices `matrix` you should pass as a parameter, maybe you use that somehow)?

Reply:
I think I had tried with a for loop (in a slightly different way) but was slower. But I'll double check again to be sure or possibly find another way